gui: Fix intro dialog labels when the prune button is toggled

hebasto opened the pull request:
On master (a6f6333ba253cda83221ee529810cacf930e413f) and on 0.19.0.1 the intro dialog with prune enabled (checkbox "Discard blocks..." is checked) provides a user with wrong info about the required disk space:

![DeepinScreenshot_bitcoin-qt_20191208112228](https://user-images.githubusercontent.com/32963518/70387510-8daab400-19ae-11ea-9338-29add9c31118.png)

Also the paragraph "If you have chosen to limit..." is missed.

---

With this PR when prune checkbox is toggled, the related text labels and the amount of required space shown are updated (previously they were only updated when the data directory was updated):
![Screenshot from 2019-12-08 11-34-53](https://user-images.githubusercontent.com/32963518/70387542-eed28780-19ae-11ea-9565-49d8a64b2f33.png)

---

This PR is an alternative to #17035.

**ryanofsky**'s [suggestion](https://github.com/bitcoin/bitcoin/pull/17035#discussion_r337594268) also has been implemented.

hebasto committed:
util: Add PruneGBtoMiB() function

This commit does not change behavior.

hebasto committed:
util: Add PruneMiBtoGB() function

Now the text of prune QCheckBox shows the space in GB instead of
thousands MiB, which is consistent with other parts of the GUI.

hebasto committed:
refactor: Replace static variable with data member

hebasto committed:
refactor: Add Intro::UpdatePruneLabels()

This is a move-only commit and it does not change behavior.

hebasto commented:
ping @ryanofsky @Sjors @promag

ryanofsky commented on src/qt/optionsmodel.h:
```diff
@@ -74,6 +74,7 @@ class OptionsModel : public QAbstractListModel
 
     /* Explicit setters */
     void SetPrune(bool prune, bool force = false);
+    void SetPrune(unsigned int prune_size_gb, bool force = false);
```
In commit "gui: Fix prune settings after intro" (5fd852990cd0e5c94e384fcf6c5fb22891534f4f)

It seems dangerous to give two functions that change different settings ("bPrune" vs "nPruneSize") the exact same name distinguished only by a `bool` `int` overload. Should give the functions different names. Might suggest `SetPruneSizeGB` and `SetPruneEnabled`.

hebasto replied:
Done.

ryanofsky commented on src/qt/bitcoin.cpp:
```diff
@@ -282,7 +282,7 @@ void BitcoinApplication::parameterSetup()
 
 void BitcoinApplication::SetPrune(bool prune, bool force) {
-     optionsModel->SetPrune(prune, force);
+    optionsModel->SetPrune(prune ? DEFAULT_PRUNE_SIZE_GB : 0, force);
```
In commit "gui: Fix prune settings after intro" (5fd852990cd0e5c94e384fcf6c5fb22891534f4f)

I can't figure out what this commit is doing. What is being fixed? What was the behavior before this commit and what is the behavior after? Description just says "If QSettings are set already, it is required to reset prune size option," which doesn't make sense to me, but I must be missing context that would make it understandable.

hebasto replied:
@ryanofsky
> I can't figure out what this commit is doing. What is being fixed?

Steps to reproduce this bug:

f25c99e735a2abc930752249687d780923e941c8 (before 5fd852990cd0e5c94e384fcf6c5fb22891534f4f)

1. run `bitcoin-qt`, set GUI option as pictured:
![Screenshot from 2019-11-12 22-07-27](https://user-images.githubusercontent.com/32963518/68707088-c1eea880-0599-11ea-939a-6a7ec19658e1.png)
and quit

2. run `bitcoin-qt -choosedatadir`, in intro set check prune:
![Screenshot from 2019-11-12 22-08-04](https://user-images.githubusercontent.com/32963518/68707261-142fc980-059a-11ea-90c7-e1f5c0e6219a.png)

Expected prune target 2 GB = 1907 MiB.
Actually, logged:
```
Prune configured to target 5722 MiB on disk for block and undo files.
```

It seems, I should split first two commits into separated PR, right?

hebasto replied:
See #17696.

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -133,9 +133,9 @@ Intro::Intro(QWidget *parent, uint64_t blockchain_size, uint64_t chain_state_siz
-    const unsigned int prune_target_gb = PruneMiBtoGB(prune_target_mib);
-    ui->prune->setText(tr("Discard blocks after verification, except most recent %1 GB (prune)").arg(prune_target_gb ? prune_target_gb : DEFAULT_PRUNE_SIZE_GB));
-    UpdatePruneLabels(prune_target_gb);
+    m_prune_target_gb = prune_target_mib ? PruneMiBtoGB(prune_target_mib) : DEFAULT_PRUNE_SIZE_GB;
```
In commit "gui: Fix intro dialog when prune is toogled" (4824eeb91e8ee7ed2e46519f56d860635e7c4818)

I can't figure out what this commit is doing. Could use description of what behavior was before and how it is changed after. Just looking at code I see it is moving away from convention of using 0 pruning size to represent no pruning, but I don't actually see how behavior is changing.

EDIT: Just noticed the new QCheckBox::toggled handler below. Maybe pull refactoring changes out of this commit to make each set of changes more obvious? Also s/toogled/toggled/

hebasto replied:
> I can't figure out what this commit is doing. Could use description of what behavior was before and how it is changed after.

The bug and fixed behavior are described in #17035 OP.

hebasto replied:
This PR description has been updated.

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -302,7 +302,10 @@ void Intro::startThread()
 
     connect(executor, &FreespaceChecker::reply, this, &Intro::setStatus);
     connect(this, &Intro::requestCheck, executor, &FreespaceChecker::check);
-    /*  make sure executor object is deleted in its own thread */
```
In commit "gui: Fix intro dialog when prune is toogled" (4824eeb91e8ee7ed2e46519f56d860635e7c4818)

Suggest not deleting this comment line assuming it is still applicable.

hebasto replied:
This line
```
connect(thread, &QThread::finished, executor, &QObject::deleteLater);
```
seems to be a standard pattern in Qt framework. Yes, the comment is still applicable, but it does not provide any additional knowledge, IMHO.

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -346,3 +329,25 @@ QString Intro::getPathToCheck()
+{
+    m_required_space_gb = m_blockchain_size;
+    QString storageRequiresMsg = tr("At least %1 GB of data will be stored in this directory, and it will grow over time.");
+    if (ui->prune->isChecked()) {
```
In commit "refactor: Add Intro::UpdatePruneLabels()" (27f716ebbadfefee8eac7e43374f080ff3dbc68a)

Why did this line change from `if (prune_target_gb)` if this commit is supposed to be move-only?

hebasto replied:
> Why did this line change from `if (prune_target_gb)` if this commit is supposed to be move-only?

You are right. It is a remnant of the previous branch. Will be fixed.

hebasto replied:
Fixed.

emilengler commented:
Thanks, Concept ACK

hebasto commented on src/qt/optionsmodel.h:
```diff
@@ -16,6 +17,9 @@ class Node;
 static constexpr unsigned short DEFAULT_GUI_PROXY_PORT = 9050;
 
+static inline uint64_t PruneGBtoMiB(unsigned int gb) { return (gb * GB_BYTES) >> 20; }
+static inline unsigned int PruneMiBtoGB(uint64_t mib) { return ((mib << 20) + GB_BYTES / 2) / GB_BYTES; }
```
Agree with rounding and types. Going to fix.

Do you think the rounding rationale should be commented? If so, what could you suggest?

ryanofsky replied:
> Do you think the rounding rationale should be commented? If so, what could you suggest?

Maybe something like "Round up when converting from configured MiB to displayed GB to avoid underestimating max disk usage. Round down in the other direction so roundtrip GB -> MiB -> GB conversion is stable."

hebasto replied:
Done.

hebasto commented:
@ryanofsky 

Thank you for your review.
All yours comments have been addressed: implemented, fixed or replied to.

Sjors commented:
When you select a volume with low disk space (e.g. a RAM disk), when the user unchecks the prune box, it automatically checks itself again.

hebasto commented:
@Sjors 
> When you select a volume with low disk space (e.g. a RAM disk), when the user unchecks the prune box, it automatically checks itself again.

It seems reasonable (the same behavior was in #17035). How should it be in your opinion?

Sjors commented:
Either the user should be allowed to go ahead anyway (maybe they'll delete some movies during IBD), or the checkbox should be disabled.

hebasto commented:
@Sjors 

Thank you for your review.

> Either the user should be allowed to go ahead anyway (maybe they'll delete some movies during IBD), or the checkbox should be disabled.

The former suggestion has been implemented in the latest push 9672ea841d7a66c765d734cfab57210baa4b074d.

ryanofsky commented:
PR description doesn't describe PR. Can you update it to say what the problem with intro dialog labels is and how the problems is fixed? Also if behavior or code organization is changing in other ways because this seems to do more than just change labels...

ryanofsky commented on src/qt/bitcoin.cpp:
```diff
@@ -282,7 +282,7 @@ void BitcoinApplication::parameterSetup()
 
 void BitcoinApplication::SetPrune(bool prune, bool force) {
-     optionsModel->SetPrune(prune, force);
+    optionsModel->SetPruneTargetGB(prune ? DEFAULT_PRUNE_TARGET_GB : 0, force);
```
In commit "gui: Fix prune in QSettings after intro" (f1ae0412d80ac7f8b30ec412564aaec782b3df0b)

I can't figure out what this commit is doing. Can the commit description be updated to say what the behavior is before and after this commit?

hebasto replied:
See #17696.

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -130,18 +131,18 @@ Intro::Intro(QWidget *parent, uint64_t blockchain_size, uint64_t chain_state_siz
         ui->prune->setChecked(true);
         ui->prune->setEnabled(false);
     }
-    ui->prune->setText(tr("Discard blocks after verification, except most recent %1 GB (prune)").arg(pruneTarget ? pruneTarget / 1000 : DEFAULT_PRUNE_TARGET_GB));
```
In commit "util: Add PruneMiBtoGB() function" (162dafaaa9d47e7d40db09040a265bccf975be20)

Maybe update commit description to note slight change in label text: now displays GB number instead of GiB number.

I think there are no other changes in behavior in this commit, or at least I don't see any other changes.

hebasto replied:
Done.



> I think there are no other changes in behavior in this commit, or at least I don't see any other changes.

Correct.

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -133,9 +133,9 @@ Intro::Intro(QWidget *parent, uint64_t blockchain_size, uint64_t chain_state_siz
-    const int prune_target_gb = PruneMiBtoGB(prune_target_mib);
-    ui->prune->setText(tr("Discard blocks after verification, except most recent %1 GB (prune)").arg(prune_target_gb ? prune_target_gb : DEFAULT_PRUNE_TARGET_GB));
-    UpdatePruneLabels(prune_target_gb);
+    m_prune_target_gb = prune_target_mib ? PruneMiBtoGB(prune_target_mib) : DEFAULT_PRUNE_TARGET_GB;
```
In commit "gui: Fix intro dialog when prune is toggled" (f07dd7c69ed7a5245f79662100c859286557b73b)

I think this should say `prune_target_mib > 1 ? ...` instead of `prune_target_mib ? ...`

If the pruning value is 1, pruning is still effectively disabled and more than 1GB will be required unless manual pruning is done.

Could also combine this check into the `if (prune_target_mib > 1)` check above.

hebasto replied:
Fixed in the latest push.

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -302,7 +302,10 @@ void Intro::startThread()
+    connect(ui->prune, &QCheckBox::toggled, [this](bool prune_checked) {
+        UpdatePruneLabels(prune_checked);
+        Q_EMIT requestCheck();
+    });
```
In commit "gui: Fix intro dialog when prune is toggled" (f07dd7c69ed7a5245f79662100c859286557b73b)

Why trigger a new disk space check here? Can you add explanatory comments? I think I understand updating the labels because the label text depends on the check state, but I don't get the disk space thing.

It would also be good to update commit description to say what behavior was before and after this change, and how it works.

hebasto replied:
> Why trigger a new disk space check here?

Reworked. No unneeded disk space check anymore.

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -255,11 +255,11 @@ void Intro::setStatus(int status, const QString &message, quint64 bytesAvailable
             freeString += " " + tr("(of %n GB needed)", "", m_required_space_gb);
             ui->freeSpace->setStyleSheet("QLabel { color: #800000 }");
-            ui->prune->setChecked(true);
+            if (!keep_prune) ui->prune->setChecked(true);
```
In commit "gui: Let user to uncheck prune on low-space dirs" (9672ea841d7a66c765d734cfab57210baa4b074d)

Can you update the commit description to say behavior was before and after this change? It seems without this code the previous commit f07dd7c69ed7a5245f79662100c859286557b73b will be broken, because you added a new space check when the checkbox is toggled, so the box would automatically toggle itself back on after the check completed? All of this is very unclear to me.

hebasto replied:
> It seems without this code the previous commit [f07dd7c](https://github.com/bitcoin/bitcoin/commit/f07dd7c69ed7a5245f79662100c859286557b73b) will be broken...

You are right. Fixed in the latest push.

ryanofsky reviewed (approved):
Partial code review ACK 9672ea841d7a66c765d734cfab57210baa4b074d

I reviewed the settings changes I'm more familiar with. Some the GUI commits went over my head though. I think they'd make more sense to me if they described the effects they have.

- [X] f2afd0705de8f0906a2ce996fa49673633caeeb1 util: Replace magics with DEFAULT_PRUNE_TARGET_GB (1/8)
- [ ] f1ae0412d80ac7f8b30ec412564aaec782b3df0b gui: Fix prune in QSettings after intro (2/8)
- [X] 162dafaaa9d47e7d40db09040a265bccf975be20 util: Add PruneMiBtoGB() function (3/8)
- [X] 0f8830b08bc19d62c995434ad47a01577eaa4b6c util: Add PruneGBtoMiB() function (4/8)
- [X] 5d8dbfa34655a9fa89a69b971451cf05356ca434 refactor: Replace static variable with data member (5/8)
- [X] 6dd682a01fd9788f4944ac4be84897ae31925f2c refactor: Add Intro::UpdatePruneLabels() (6/8)
- [ ] f07dd7c69ed7a5245f79662100c859286557b73b gui: Fix intro dialog when prune is toggled (7/8)
- [ ] 9672ea841d7a66c765d734cfab57210baa4b074d gui: Let user to uncheck prune on low-space dirs (8/8)

hebasto commented:
@ryanofsky 

> PR description doesn't describe PR. Can you update it to say what the problem with intro dialog labels is and how the problems is fixed?

The OP has been updated.

hebasto commented:
@ryanofsky 
> In commit "gui: Fix prune in QSettings after intro" ([f1ae041](https://github.com/bitcoin/bitcoin/commit/f1ae0412d80ac7f8b30ec412564aaec782b3df0b))
> 
> I can't figure out what this commit is doing. Can the commit description be updated to say what the behavior is before and after this commit?

The first two commits has been split of into #17696, as they are dedicated to an orthogonal bug.
The related conversations are marked as resolved here.

hebasto committed:
gui: Add Intro::UpdateFreeSpaceLabel()

If a new custom datadir has enough free space, the prune checkbox gets
unchecked, unless -prune=NNN command-line option is provided.

hebasto committed:
gui: Make Intro consistent with prune checkbox

When prune checkbox is toggled, the related text labels and the amount
of required space shown are updated (previously they were only updated
when the data directory was updated).

hebasto commented:
@ryanofsky 
Thank you for your review. All your comments have been addressed.

Would you mind re-reviewing this PR?

hebasto commented:
Rebased after #17696 has been merged.

ryanofsky reviewed (approved):
Code review ACK 3f28e1025a97f29f5265d75cdca0cd9e4330a087. This looks good. Description of the PR is very clear, and the commits are well organized and mostly a breeze to review. There's a lot going on in the final commit, and I left some comments there to help clarify it, but feel free to ignore my suggestions if they don't make sense

hebasto commented on src/qt/intro.cpp:
```diff
@@ -251,24 +251,29 @@ void Intro::setStatus(int status, const QString &message, quint64 bytesAvailable
-        } else if (bytesAvailable / GB_BYTES - m_required_space_gb < 10) {
-            freeString += " " + tr("(%n GB needed for full chain)", "", m_required_space_gb);
-            ui->freeSpace->setStyleSheet("QLabel { color: #999900 }");
-            ui->prune->setChecked(true);
```
> The difference with this PR is just that the checkbox gets unchecked when a new datadir is selected, where previously it would never get unchecked?

Steps to reproduce:
1) the default datadir not having enough free space causes the prune checkbox gets checked (both on master and with this PR)
2) a user chooses a custom datadir with enough free space:
  - on master the prune checkbox remains checked
  - with this PR the prune checkbox gets unchecked

ryanofsky replied:
In commit "gui: Add Intro::UpdateFreeSpaceLabel()" (a208b68b6f30f932d82e0d625bb230f9c1d5ab72)

> > The difference with this PR is just that the checkbox gets unchecked when a new datadir is selected, where previously it would never get unchecked?
> 
> Steps to reproduce: [...]

Thanks for the description. It sounds like the answer is yes.

hebasto commented on src/qt/intro.cpp:
```diff
@@ -128,14 +126,22 @@ Intro::Intro(QWidget *parent, uint64_t blockchain_size, uint64_t chain_state_siz
+    });
+
+    int64_t prune_target_mib = gArgs.GetArg("-prune", 0);
+    if (prune_target_mib * 1024 * 1024 >= MIN_DISK_SPACE_FOR_BLOCK_FILES) {
```
> The two versions are equivalent because of [other code forbidding values from 2 to 549](https://github.com/bitcoin/bitcoin/blob/6196e930018181301b5972842ae384ea4288ff34/src/init.cpp#L1061-L1063)).

That is not the case.
The code you mentioned is called much later:
https://github.com/bitcoin/bitcoin/blob/e7f84503571c171a7e6728cd2d77dd4103bd7a6f/src/qt/bitcoin.cpp#L581
after
https://github.com/bitcoin/bitcoin/blob/e7f84503571c171a7e6728cd2d77dd4103bd7a6f/src/qt/bitcoin.cpp#L491

It implies that forbidding values from 2 to 549 are accepted and
```
$ bitcoin-qt -choosedatadir -prune=42
```
works as `prune=1`.

This PR ignores prune forbidding values from 2 to 549.

Move it to a separate _non_-refactoring commit?

hebasto replied:
> Note: I found the change on this line confusing, and might suggest reverting...

Reverted.

hebasto commented on src/qt/intro.h:
```diff
@@ -72,11 +72,12 @@ private Q_SLOTS:
     //! Total required space (in GB) depending on user choice (prune or not prune).
     int64_t m_required_space_gb{0};
     uint64_t m_bytes_available{0};
+    int64_t m_prune_target_gb{0};
```
> ... having this function would still simplify the `Intro::Intro` code and allow declaring the member const.

Implemented.

hebasto commented on src/qt/intro.cpp:
```diff
@@ -128,14 +126,22 @@ Intro::Intro(QWidget *parent, uint64_t blockchain_size, uint64_t chain_state_siz
+    connect(ui->prune, &QCheckBox::toggled, [this](bool prune_checked) {
+        UpdatePruneLabels(prune_checked);
+        UpdateFreeSpaceLabel();
+    });
```
> Current commit description is vague.

Fixed.

hebasto commented on src/qt/intro.cpp:
```diff
@@ -138,21 +135,21 @@ Intro::Intro(QWidget *parent, uint64_t blockchain_size, uint64_t chain_state_siz
     const int prune_target_gb = PruneMiBtoGB(prune_target_mib);
     ui->prune->setText(tr("Discard blocks after verification, except most recent %1 GB (prune)").arg(prune_target_gb ? prune_target_gb : DEFAULT_PRUNE_TARGET_GB));
-    requiredSpace = m_blockchain_size;
+    m_required_space_gb = m_blockchain_size;
```
The code is reorganized to make assigning values to `m_required_space_gb` variable more clear.

hebasto commented:
Updated 3f28e1025a97f29f5265d75cdca0cd9e4330a087 -> 91cc02dbeb026f5f029b7368c25739c754f19667 ([pr17453.10](https://github.com/hebasto/bitcoin/commits/pr17453.10) -> [pr17453.11](https://github.com/hebasto/bitcoin/commits/pr17453.11), [compare](https://github.com/hebasto/bitcoin/compare/pr17453.10..pr17453.11)).

@ryanofsky 
Thank you for your review. Your comments have been addressed.

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -138,21 +135,20 @@ Intro::Intro(QWidget *parent, uint64_t blockchain_size, uint64_t chain_state_siz
     }
     const int prune_target_gb = PruneMiBtoGB(prune_target_mib);
     ui->prune->setText(tr("Discard blocks after verification, except most recent %1 GB (prune)").arg(prune_target_gb ? prune_target_gb : DEFAULT_PRUNE_TARGET_GB));
-    requiredSpace = m_blockchain_size;
```
In commit "refactor: Replace static variable with data member" (02c6060ba65c160dcd336b57752ce304ecd4ca9f)

This line doesn't seem to be replaced by anything. It looks like `m_required_space_gb` will now be set to 0 instead of `m_blockchain_size` when `prune_target_gb` is greater than `m_blockchain_size_gb`.

It seems like this should set `m_required_space_gb = m_blockchain_size_gb + m_chain_state_size_gb` when `prune_target_gb > m_blockchain_size_gb`

hebasto replied:
Yes, this is a bug, and it is easy to reproduce it. Going to fix it.

ryanofsky commented on src/qt/intro.cpp:
```diff
@@ -345,3 +329,24 @@ QString Intro::getPathToCheck()
+{
+    QString storageRequiresMsg = tr("At least %1 GB of data will be stored in this directory, and it will grow over time.");
+    if (prune_target_gb) {
+        if (prune_target_gb <= m_blockchain_size_gb) {
```
In commit "refactor: Add Intro::UpdatePruneLabels()" (9759ccbb40a1ce333d8ae3b6c3ed82520a948b76)

It seems like this should set `m_required_space_gb = m_blockchain_size_gb + m_chain_state_size_gb` in the case where `prune_target_gb > m_blockchain_size_gb` instead of not updating `m_required_space_gb`

ryanofsky reviewed (approved):
Conditional code review ACK 91cc02dbeb026f5f029b7368c25739c754f19667 if questions below about `m_required_space_gb` being left 0 when prune size is greater than blockchain size are addressed.

Changes since last review:

- More variables renamed, made signed and given `_gb` suffixes
- Possible bug added in commit "refactor: Replace static variable with data member" could let `m_required_space_gb` be 0
- Prune box unchecking code in commit "gui: Add Intro::UpdateFreeSpaceLabel()" is fixed to check enabled status and not rely on m_required_space_gb. Commit description is also updated
- Label updating changes in commit "gui: Make Intro consistent with prune checkbox" have some refactoring and go back to treating values 2-549 as pruning to simplify checks and avoid changing behavior

hebasto commented:
Updated 91cc02dbeb026f5f029b7368c25739c754f19667 -> 95761b471e22c3441035901bc0113004d9a53ec4 ([pr17453.11](https://github.com/hebasto/bitcoin/commits/pr17453.11) -> [pr17453.12](https://github.com/hebasto/bitcoin/commits/pr17453.12), [compare](https://github.com/hebasto/bitcoin/compare/pr17453.11..pr17453.12)).

@ryanofsky 
Thank you for your review. The bug you [mentioned](https://github.com/bitcoin/bitcoin/pull/17453#discussion_r365947039) has been fixed.

hebasto commented:
Updated 91cc02dbeb026f5f029b7368c25739c754f19667 -> 4f7127d1e3a51f0f55d42a08439c516dcc8d1a26 ([pr17453.11](https://github.com/hebasto/bitcoin/commits/pr17453.11) -> [pr17453.12](https://github.com/hebasto/bitcoin/commits/pr17453.12), [compare](https://github.com/hebasto/bitcoin/compare/pr17453.11..pr17453.12)).

@ryanofsky 
Thank you for your review. The bug you [mentioned](https://github.com/bitcoin/bitcoin/pull/17453#discussion_r365947039) has been fixed.

ryanofsky reviewed (approved):
Code review ACK 4f7127d1e3a51f0f55d42a08439c516dcc8d1a26. It seems like there are a few visible changes here:

- Main change is amount of required disk space display now being updated when PR is checked and unchecked. This change is mentioned in the PR description and implemented in commit "gui: Make Intro consistent with prune checkbox"

- Another change is required space for pruning now being converted correctly to GB instead of GiB in commit "util: Add PruneMiBtoGB() function"

- Another change is that the pruning checkbox will automatically get unchecked if a newly selected datadir has enough free space to hold the chainstate and blocks in commit "gui: Add Intro::UpdateFreeSpaceLabel()"

- Another change is that "Discard blocks after verification, except most recent %1 GB" label now shows 2GB instead of 0GB if bitcoin is started with `-prune=1` and does the MiB to GB conversion more correctly in other cases where `-prune=` is set, using the right conversion factor and rounding up instead of down

emilengler commented on src/qt/intro.cpp:
```diff
@@ -109,52 +107,49 @@ void FreespaceChecker::check()
     Q_EMIT reply(replyStatus, replyMessage, freeBytesAvailable);
 }
 
+namespace {
```
Why do we need a namespace here?

hebasto replied:
[Use an unnamed (anonymous) namespace for all internal/nonexported entities](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#sf22-use-an-unnamed-anonymous-namespace-for-all-internalnonexported-entities)

emilengler replied:
Why has it no intends?

hebasto replied:
> Why has it no intends?

Sorry. "intends" or "indents"?

emilengler replied:
indents, sorry

hebasto replied:
It follows our [Coding Style](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c). Also see [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy).

emilengler commented:
I tested it and it works and the code also looks OK to me.
Please squash and I will ACK.
This should get into 0.19.1 btw. Maybe it's worth talking about it on the next meeting

hebasto commented:
@emilengler 
> Please squash...

It seems it could make reviewing a bit harder for other reviewers, no?

emilengler commented:
> It seems it could make reviewing a bit harder for other reviewers, no?

If the commit message matters, put them into the longer description. But I don't think it is worth looking at the diffs from commit to commit. It is probably the easiest to just look at the diff from HEAD to master

sipa commented:
@emilengler Even if there are multiple commits, you're free as a reviewer to look at the entire diff at once. The decision whether to break things up is mostly up to the author, with the caveat that each commit must make sense on its own, and ideally, the code compiles and passes tests for every individual step. I don't see a reason here to request squashing - that's usually only done when a PR contains "fixups" as a result of review that really should be integrated into the body.

emilengler commented:
ACK

Sjors commented:
tACK 4f7127d1e3a51f0f55d42a08439c516dcc8d1a26

jonasschnelli reviewed (approved):
utACK 4f7127d1e3a51f0f55d42a08439c516dcc8d1a26
